native/cmake/predefine.cmake

```diff
@@ -232,8 +237,99 @@ function(cc_inspect_values)
     endforeach()
 endfunction()
 
+function(cc_parse_cfg_include_files cfg_file output_var)
+    file(STRINGS ${cfg_file} my_lines)
+    set(include_pattern "^%include\\s*\"([^\"]*)\"\\s*")
+    set(include_files "")
+    foreach(line ${my_lines})
+        if(line MATCHES ${include_pattern})
+            # keep syncing with SWIG_ARGS
+            set(include_file ${CMAKE_CURRENT_LIST_DIR}/cocos/${CMAKE_MATCH_1})
+            set(include_file2 ${CMAKE_CURRENT_LIST_DIR}/${CMAKE_MATCH_1})
+            if(EXISTS ${include_file})
+                list(APPEND include_files ${include_file})
+            elseif(EXISTS ${include_file2})
+                list(APPEND include_files ${include_file2})
+            else()
+                message(FATAL_ERROR "%include ${include_file}:  file not found")
+            endif()
+        endif()
+    endforeach()
+    set(${output_var} ${include_files} PARENT_SCOPE)
+endfunction()
+
+function(cc_gen_swig_files cfg_directory output_dir)
+    file(MAKE_DIRECTORY "${output_dir}")
+    if(${CMAKE_HOST_SYSTEM_NAME} MATCHES "Windows")
+        set(SWIG_DIR ${EXTERNAL_ROOT}/win64/bin/swig) 
+        set(SWIG_EXEC ${SWIG_DIR}/bin/swig.exe)
+    elseif(${CMAKE_HOST_SYSTEM_NAME} MATCHES "Darwin")
+        set(SWIG_DIR ${EXTERNAL_ROOT}/mac/bin/swig) 
+        set(SWIG_EXEC ${SWIG_DIR}/bin/swig)
+    elseif(${CMAKE_HOST_SYSTEM_NAME} STREQUAL "Linux")
+        set(SWIG_DIR ${EXTERNAL_ROOT}/linux/bin/swig) 
+        set(SWIG_EXEC ${SWIG_DIR}/bin/swig)
+    else()
+        message(FATAL_ERROR "swig is not supported on current platform!")
+    endif()
+
+    set(SWIG_ARGS
+        -c++
+        -cocos
+        -fvirtual
+        -noexcept
+        -cpperraswarn
+        -D__clang__
+        -Dfinal= 
+        -DCC_PLATFORM=3
+        -Dconstexpr=const
+        -DCC_PLATFORM_ANDROID=3
+        -I${SWIG_DIR}/share/swig/4.1.0/javascript/cocos
+        -I${SWIG_DIR}/share/swig/4.1.0
+        -I${CMAKE_CURRENT_LIST_DIR}/
+        -I${CMAKE_CURRENT_LIST_DIR}/cocos
+        -o
+    )
+    file(GLOB cfg_files ${cfg_directory}/*.i)
+
+    list(FILTER cfg_files EXCLUDE REGEX ".*template.*")
+    foreach(cfg ${cfg_files})
+
+        set(dep_files)
+        get_filename_component(mod_name ${cfg} NAME_WE)
+        file(MAKE_DIRECTORY ${output_dir}/temp)
+        set(output_file_tmp ${output_dir}/temp/jsb_${mod_name}_auto.cpp)
+        set(output_file ${output_dir}/jsb_${mod_name}_auto.cpp)
+
+        set(output_hfile_tmp ${output_dir}/temp/jsb_${mod_name}_auto.h)
+        set(output_hfile ${output_dir}/jsb_${mod_name}_auto.h)
 
+        cc_parse_cfg_include_files(${cfg} dep_files)
 
+        add_custom_command(
+            OUTPUT
+                ${output_hfile}
+                ${output_file}
+            COMMAND ${CMAKE_COMMAND} -E echo "Running swig with config file ${cfg} ..."
+            COMMAND ${SWIG_EXEC} ${SWIG_ARGS}
```

Comment:
Could we invoke `node genbindings.js` rather than `calling swig executable file`?

Reply:
This might introduce a dependency on Node.